src/fairseq2/metrics.py

```diff
@@ -158,24 +158,218 @@ def sync_and_compute_metrics(*bags: MetricBag) -> Optional[Dict[str, Any]]:
-                fmt_value = bag.format_metric_value(name, value)
-                if fmt_value is not None:
-                    break
+    if postfix:
+        s += postfix
+
+    return s
+
+
+format_as_seconds = partial(format_as_int, postfix="s")
+"""Format metric ``value`` as duration in seconds."""
+
+
+def format_as_float(
+    value: Any, *, decimal: Optional[int] = None, postfix: Optional[str] = None
+) -> str:
+    """Format metric ``value`` as float."""
+    if decimal:
+        s = f"{float(value):,.{decimal}f}"
+    else:
+        s = f"{float(value):,}"
+
+    if postfix:
+        s += postfix
+
+    return s
+
+
+format_as_loss = partial(format_as_float, decimal=3)
+"""Format metric ``value`` as training loss."""
+
+
+_metric_formatters: Dict[str, Tuple[str, Callable[[Any], str]]] = {
+    # fmt: off
+    "batch_size":          ("Batch Size",                format_as_int),
+    "elapsed_time":        ("Elapsed Time",              format_as_seconds),
+    "elements_per_batch":  ("Elements per Batch",        format_as_int),
+    "elements_per_second": ("Elements per Second",       format_as_int),
+    "entropy_loss":        ("Entropy Loss",              format_as_loss),
+    "grad_scale":          ("Grad Scale",                format_as_float),
+    "loss":                ("Loss",                      format_as_loss),
+    "lr":                  ("Learning Rate",             format_as_float),
+    "num_source_elements": ("Number of Source Elements", format_as_int),
+    "num_target_elements": ("Number of Target Elements", format_as_int),
+    "wall_time":           ("Wall Time",                 format_as_seconds),
+    # fmt: on
+}
+
+
+def register_metric_formatter(
+    name: str, display_name: str, formatter: Callable[[Any], str]
+) -> None:
+    """Register a string formatter for the specified metric.
+
+    :param name:
+        The name of the metric.
+    :param display_name:
+        The display name of the metric.
+    :param formatter:
+        The formatter to convert a metric value to its string representation.
+    """
+    if name in _metric_formatters:
+        raise ValueError(
+            f"`name` must be a unique metric name, but '{name}' is already registered."
+        )
+
+    _metric_formatters[name] = (display_name, formatter)
+
+
+class MetricRecorder(ABC):
+    """Records metric values."""
+
+    @abstractmethod
+    def record_metrics(
+        self,
+        run: str,
+        values: Dict[str, Any],
+        step_nr: int,
+        *,
+        flush: Optional[bool] = False,
+    ) -> None:
+        """Record ``values``.
+
+        :param run:
+            The name of the run (e.g. 'train', 'eval').
+        :param values:
+            The metric values.
+        :param step_nr:
+            The number of the run step.
+        :param flush:
+            If ``True``, flushes any intermediate buffer after recording.
+        """
+
+    def close(self) -> None:
+        """Close the recorder."""
```

Comment:
nit: shouldnt this have a `pass` statement or something like that

Reply:
It does not necessarily have to have, syntactically this is correct, but to increase the readability I added a "redundant" `pass` in #287.

Reply:
Thank you for the clarification!